### PR TITLE
Introduce Mapper

### PR DESCRIPTION
**Pull request by dongreenberg**
Expand mapping, replicating, and load blancing functionality via a built-in Module called rh.Mapper. Mapper
takes a function and creates replicas of it and its envs, so when mapper.map(*args) is called, it can simply
round-robin call the replicas. It supports user-added replicas as well.

### CODE REVIEW

**Comment by dongreenberg** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #327** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/327?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/run-house/runhouse/327?utm_source=stack-comment).

File: runhouse/resources/functionals/mapper.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+from runhouse.resources.functions import function, Function
+
+from runhouse.resources.module import Module
+
+logger = logging.getLogger(__name__)
+
+
+class Mapper(Module):
+    def __init__(self, module, method, num_replicas=-1, replicas=None, **kwargs):
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:13]: mypy types? `replicas: Optional[List] = None`

also can we call it different?

File: runhouse/resources/functionals/mapper.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        self._auto_replicas.extend(self.module.replicate(num_replicas))
+
+    def increment_counter(self):
+        self._last_called += 1
+        if self._last_called >= len(self.replicas):
+            self._last_called = 0
+        return self._last_called
+
+    @staticmethod
+    def _call_method_on_replica(replica, method, args, kwargs):
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:49]: pls more mypy types

File: runhouse/resources/module.py [hunk 3]
```diff
@@ -603,6 +604,48 @@ def refresh(self):
         else:
             return self
 
+    def replicate(self, num_replicas=1, names="auto", envs="auto"):
```
Comment (rohinb2) [inline runhouse/resources/module.py:607]: also should check `if envs != None`, `assert len(envs) == num_replicas`

Comment (rohinb2) [inline runhouse/resources/module.py:607]: does `num_replicas` include the current?

File: runhouse/resources/module.py [hunk 4]
```diff
@@ -603,6 +604,48 @@ def refresh(self):
...
+            else:
+                raise ValueError(
+                    "names must be a list of names or 'auto' to auto-generate names."
+                )
+
+            if isinstance(envs, list):
+                env = _get_env_from(envs[i])
+            elif envs == "auto":
+                env_conf = self.env.config_for_rns
+                env_conf["name"] = f"{self.env.name}_{i}"
```
Comment (rohinb2) [inline runhouse/resources/module.py:633]: i like `{self.env.name}_rh_replica_{i}`

File: runhouse/resources/functionals/mapper.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    def __init__(self, module, method, num_replicas=-1, replicas=None, **kwargs):
+        super().__init__(**kwargs)
+        self.module = module
+        self.method = method
+        self.num_replicas = num_replicas
+        self._auto_replicas = []
+        self._user_replicas = replicas or []
+        self._last_called = 0
+        if self.num_replicas > len(self.replicas) and self.num_replicas > 0:
+            self._add_auto_replicas(self.num_replicas - len(self.replicas))
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:22]: i'd say get rid of this helper and do it here, more clear

File: runhouse/resources/functionals/mapper.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    @property
+    def replicas(self):
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:25]: i like `replicated_modules`, way more clear that these are all `Modules` we're dealing with

Comment (rohinb2) [reply runhouse/resources/functionals/mapper.py:25]: i'd propagate this name change everywhere

File: runhouse/resources/functionals/mapper.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        )
+
+    def call(self, *args, **kwargs):
+        """Call the function or method on a single replica.
+
+        Example:
+            >>> def local_sum(arg1, arg2, arg3):
+            >>>     return arg1 + arg2 + arg3
+            >>>
+            >>> remote_fn = rh.function(local_fn).to(gpu)
```
[the remark this answers is not in the document]
Comment (dongreenberg) [reply runhouse/resources/functionals/mapper.py:108]: 🤦‍♂️

File: runhouse/resources/functionals/mapper.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        Example:
+            >>> def local_sum(arg1, arg2, arg3):
+            >>>     return arg1 + arg2 + arg3
+            >>>
+            >>> remote_fn = rh.function(local_sum).to(my_cluster)
+            >>> mapper = rh.mapper(remote_fn, num_replicas=2)
+            >>> arg_list = [(1,2), (3, 4)]
+            >>> # runs the function twice, once with args (1, 2) and once with args (3, 4)
+            >>> remote_fn.starmap(arg_list)
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:88]: this should be `mapper.starmap` right?

File: runhouse/resources/module.py (hunk 3 above)
Comment (dongreenberg) [reply runhouse/resources/module.py]: Yes. We can switch it or make it an option if a users want that.

File: runhouse/resources/module.py (hunk 4 above)
Comment (dongreenberg) [reply runhouse/resources/module.py]: Met you in the middle at `{self.env.name}_replica_{i}` as discussed.

Comment (dongreenberg) [reply runhouse/resources/module.py]: I love that your urge for explicit naming goes as far as "reminder, this runhouse env was created with runhouse"

File: runhouse/resources/functionals/mapper.py (hunk 5 above)
Comment (dongreenberg) [reply runhouse/resources/functionals/mapper.py:22]: Oh that reminds me, I meant to do it in add_replicas too.

File: runhouse/resources/functionals/mapper.py (hunk 6 above)
Comment (dongreenberg) [reply runhouse/resources/functionals/mapper.py:25]: I'm leaving it to be consistent with other uses of `replicas`, notably ray serve

**Comment by dongreenberg** [comment]
### Merge activity

* **Jan 11, 8:37 PM**: @dongreenberg started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/run-house/runhouse/327).

**Pull request merged**